File: 003.longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @file: Describe the file
 * @author: tsingwong
 * @Date: 2019-02-12 19:10:54
 * @Last Modified by: tsingwong
 * @Last Modified time: 2019-02-12 19:12:05
 */

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    if (!s) {
        return 0;
    }
    let ans = 1;

    for (let i = 0; i < s.length; i++) {
        let count = 1;
        let obj = {};
        obj[s[i]] = s[i];
        for (let j = i + 1; j < s.length; j++) {
            if (obj[s[j]]) {
                break;
            } else {
                count++;
                obj[s[j]] = s[j];
            }
        }
        ans = ans >= count ? ans : count;
    }

    return ans;
};


/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring2 = function(s) {
    let lastOccurred = {};
    let start = 0;
    let maxLength = 0;
    for (let i = 0; i < s.length; i++) {
        if (lastOccurred[s[i]] >= start) {
            start = lastOccurred[s[i]] + 1;
        }
        maxLength = Math.max(i - start + 1, maxLength);
        lastOccurred[s[i]] = i;
    }
    return maxLength;
};

File: 003.longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.go
/**
 * @file: Describe the file
 * @author: tsingwong
 * @Date: 2019-02-12 19:12:17
 * @Last Modified by: tsingwong
 * @Last Modified time: 2019-02-12 19:12:39
 */
func lengthOfLongestSubstring(s string) int {
    // 记录每个字符出现的最后位置
	lastOccureed := make(map[byte]int)
	start := 0
	maxLength := 0

	for i, ch := range []byte(s) {
		
		if lastI, ok := lastOccureed[ch]; ok && lastI >= start {
			start = lastI + 1
		}
		if i - start + 1 > maxLength {
			maxLength = i - start + 1
		}
		lastOccureed[ch] = i
	}
    return maxLength
}
